src/runtime_src/core/tools/xbutil2/SubCmdStatus.cpp

```diff
@@ -0,0 +1,605 @@
+/**
+ * Copyright (C) 2019-2020 Xilinx, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may
+ * not use this file except in compliance with the License. A copy of the
+ * License is located at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+// ------ I N C L U D E   F I L E S -------------------------------------------
+// Local - Include Files
+#include "SubCmdStatus.h"
+#include "tools/common/XBUtilities.h"
+namespace XBU = XBUtilities;
+
+#include "core/include/xrt.h"
+
+#include "core/common/system.h"
+#include "core/common/device.h"
+#include "core/common/error.h"
+#include "core/common/xrt_profiling.h"
+#include "core/include/xcl_axi_checker_codes.h"
+
+// 3rd Party Library - Include Files
+#include <boost/program_options.hpp>
+#include <boost/format.hpp>
+namespace po = boost::program_options;
+
+// System - Include Files
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <iomanip>
+
+
+// ----- C L A S S   M E T H O D S -------------------------------------------
+
+const int8_t maxDebugIpType = TRACE_S2MM_FULL+1;
+
+bool debugIpOpt[maxDebugIpType];
+uint64_t debugIpNum[maxDebugIpType];
+static const char* debugIpNames[maxDebugIpType] = {
+  "unknown",
+  "LAPC",
+  "ILA",
+  "AIM",
+  "TraceFunnel",
+  "TraceFifoLite",
+  "TraceFifoFull",
+  "AM",
+  "ASM",
+  "AxiStreamProtocolChecker",
+  "TS2MM",
+  "AxiDMA",
+  "TS2MMFull"
+};
+
+
+uint32_t getDebugIpData(debug_ip_layout* map, int type,
+    std::vector<uint64_t> *baseAddress, std::vector<std::string> * portNames) 
+{
+  uint32_t count = 0;
+  for(uint64_t i = 0; i < map->m_count; i++) {
+    if(map->m_debug_ip_data[i].m_type != type)
+      continue;
+
+    if (baseAddress)
+      baseAddress->push_back(map->m_debug_ip_data[i].m_base_address);
+    if(portNames) {
+      std::string portName;
+      // Fill up string with 128 characters (padded with null characters)
+      portName.assign(map->m_debug_ip_data[i].m_name, 128);
+      // Strip away any extraneous null characters
+      portName.assign(portName.c_str());
+      portNames->push_back(portName);
+    }
+    ++count;
+  }
+  return count;
+}
```

Comment:
What is the difference between this function and the one of same name in shim.cpp?
I understand that this one takes the debug_ip_layout as argument, but rest seems the same.  If so, please refactor appropriately.

Reply:
You are right, this function is similar to what is in shim.cpp. The reason I had to add same method at both places is that "portnames" are used in the caller methods here. One solution could have been to populate portnames in the methods in shim.cpp and return using xclDebugReadIPStatus. But the prototype for xclDebugReadIPStatus is fixed for all the SHIM. Changing prototype of this API may not be justified at this point of release.
Another possible solution could be to add "getDebugIPData" as an API, so that it can be called from SubCmdStatus instead of having a copy of the code. Then, we need to add empty stubs for all the SHIMs. 
I also want to mention here that for PCIeLinux, Edge flow, the implementation has the same issue.

Reply:
It would be best to consolidate meta data parsing into some central place.  Neither this file, nor shim.cpp are the right place.  Take a look at core/common/xclbin_parser.{h,cpp} and see if you can consolidate it there.